src/meili/geometry_helpers.cc

```diff
@@ -0,0 +1,101 @@
+#include "meili/geometry_helpers.h"
+
+#include <valhalla/midgard/constants.h>
+#include <valhalla/midgard/distanceapproximator.h>
+#include <valhalla/midgard/encoded.h>
+#include <valhalla/midgard/pointll.h>
+#include <valhalla/midgard/util.h>
+
+using namespace valhalla;
+using namespace valhalla::meili;
+using namespace valhalla::midgard;
+
+namespace valhalla {
+namespace meili {
+namespace helpers {
+
+// snapped point, squared distance, segment index, offset
+std::tuple<PointLL, float, typename std::vector<PointLL>::size_type, float>
+Project(const projector_t& p, Shape7Decoder<midgard::PointLL>& shape, float snap_distance) {
+  PointLL first_point(shape.pop());
+  auto closest_point = first_point;
+  auto closest_segment_point = first_point;
+  float closest_distance = p.approx.DistanceSquared(closest_segment_point);
+  size_t closest_segment = 0;
+  float closest_partial_length = 0.f;
+  float total_length = 0.f;
```

Comment:
are you sure you dont want to at least try changing those floats to double and the one on line 57 just to at least see what it does with the test case? it seems pretty low effort to just look, in my opinion

Reply:
ive just tried updating this method on master to use `double` and all the tests pass so i think it would be pretty reasonable to also fix that in this PR. if you disagree i can just PR what ive done separately. what do you think?

Reply:
for reference

```diff
diff --git a/valhalla/meili/geometry_helpers.h b/valhalla/meili/geometry_helpers.h
index 39e0235fc..f90da6c7b 100644
--- a/valhalla/meili/geometry_helpers.h
+++ b/valhalla/meili/geometry_helpers.h
@@ -20,14 +20,14 @@ namespace helpers {
 inline std::tuple<midgard::PointLL, float, typename std::vector<midgard::PointLL>::size_type, float>
 Project(const midgard::projector_t& p,
         midgard::Shape7Decoder<midgard::PointLL>& shape,
-        float snap_distance = 0.f) {
+        double snap_distance = 0) {
   midgard::PointLL first_point(shape.pop());
   auto closest_point = first_point;
   auto closest_segment_point = first_point;
-  float closest_distance = p.approx.DistanceSquared(closest_point);
+  double closest_distance = p.approx.DistanceSquared(closest_point);
   size_t closest_segment = 0;
-  float closest_partial_length = 0.f;
-  float total_length = 0.f;
+  double closest_partial_length = 0;
+  double total_length = 0;
 
   // for each segment
   auto u = first_point;
@@ -57,23 +57,23 @@ Project(const midgard::projector_t& p,
   // the closest point on LineString to the given Point, as a fraction
   // of total 2d line length.
   closest_partial_length += closest_segment_point.Distance(closest_point);
-  float offset = total_length > 0.f ? static_cast<float>(closest_partial_length / total_length) : 0.f;
-  offset = std::max(0.f, std::min(offset, 1.f));
+  double offset = total_length > 0 ? static_cast<float>(closest_partial_length / total_length) : 0;
+  offset = std::max(0.0, std::min(offset, 1.0));
 
   // Snap to vertices if it's close
   if (total_length * offset <= snap_distance) {
     closest_point = first_point;
     closest_distance = p.approx.DistanceSquared(closest_point);
     closest_segment = 0;
-    offset = 0.f;
-  } else if (total_length * (1.f - offset) <= snap_distance) {
+    offset = 0;
+  } else if (total_length * (1 - offset) <= snap_distance) {
     closest_point = u;
     closest_distance = p.approx.DistanceSquared(closest_point);
     closest_segment = i - 1;
     offset = 1.f;
   }
 
-  return std::make_tuple(std::move(closest_point), closest_distance, closest_segment, offset);
+  return std::make_tuple(std::move(closest_point), closest_distance, closest_segment, static_cast<float>(offset));
 }
 
 } // namespace helpers
```

Reply:
Like I said, the issue can occur with double's too.  It just so happens that double's have enough precision to handle this case.  I'd be happy to switch my logic over to double's.  But I will keep my snapping logic - just adapt it to double's.

Reply:
Ok sure.  But like I said double's will have this problem too.  Switching to double might fix _this_ problem, but isn't the correct long term fix.

Reply:
yeah im not arguing that switching to double is the long term fix, im saying why not also do that clean up while we are here. it sucks that it invalidates the test but id say just keep the test and the changes but also update the function to use doubles

Reply:
Sure, I can move things to double's.